### PR TITLE
feat: use oras-credential-go to do auth

### LINKED ISSUE #931: integrated with oras-credential-go
### What is the version of your ORAS CLI

1.0.0

### What would you like to be added?

Use the implementation of oras-project/oras-credentials-go with docker/cli equivalence to login and logout.

### Why is this needed for ORAS?

CLIs like [oras](https://github.com/oras-project/oras) and [notation](https://github.com/notaryproject/notation) still rely on docker config files for credentials since it is convenient. Both CLIs re-invent the wheels at some levels, and I guess more CLIs like Helm will do the same. In addition, the issue https://github.com/oras-project/oras/issues/414 points out there might be potential bugs when sharing the same docker config with various tools with different docker CLI library versions.

### Are you willing to submit PRs to contribute to this feature?

- [X] Yes, I am willing to implement it.

### PR DESCRIPTION
**Pull request by qweeah**
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #931 

**Please check the following list**:
- [x]  Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [ ]  Does this change require a documentation update?
- [ ]  Does this introduce breaking changes that would require an announcement or bumping the major version?
- [ ]  Do all new files have an appropriate license header?

### CODE REVIEW

File: cmd/oras/root/login.go [hunk 1]
```diff
@@ -103,28 +104,16 @@ func runLogin(ctx context.Context, opts loginOptions) (err error) {
...
 	if err != nil {
 		return err
 	}
-	if err = remote.Ping(ctx); err != nil {
-		return err
-	}
-
-	// Store the validated credential
-	store, err := credential.NewStore(opts.Configs...)
+	// Ping to ensure credential is valid
```
Comment (Wwwsylvia) [inline cmd/oras/root/login.go:111]: nit: Is this comment still necessary?

**Review by Wwwsylvia — APPROVED** [review]
LGTM

File: internal/credential/store.go [hunk 2]
```diff
@@ -16,114 +16,25 @@ limitations under the License.
...
-// Store provides credential CRUD operations.
-type Store struct {
-	configs []*configfile.ConfigFile
-}
-
-// NewStore generates a store based on the passed in config file path.
-func NewStore(configPaths ...string) (*Store, error) {
+// NewStore generates a store based on the passed-in config file paths.
+func NewStore(configPaths ...string) (store credentials.Store, err error) {
+	so := credentials.StoreOptions{AllowPlaintextPut: true}
```
Comment (shizhMSFT) [inline internal/credential/store.go:24]: `opts` is a better name than `so`.

File: internal/credential/store.go [hunk 3]
```diff
@@ -16,114 +16,25 @@ limitations under the License.
...
 
-// Store provides credential CRUD operations.
-type Store struct {
-	configs []*configfile.ConfigFile
-}
-
-// NewStore generates a store based on the passed in config file path.
-func NewStore(configPaths ...string) (*Store, error) {
+// NewStore generates a store based on the passed-in config file paths.
+func NewStore(configPaths ...string) (store credentials.Store, err error) {
```
Comment (shizhMSFT) [inline internal/credential/store.go:23]: Is there any particular reason to use _named_ parameters? Please avoid them unless necessary.

File: internal/credential/store.go [hunk 4]
```diff
@@ -16,114 +16,25 @@ limitations under the License.
...
-	"context"
-	"fmt"
-	"os"
-
-	"github.com/docker/cli/cli/config"
-	"github.com/docker/cli/cli/config/configfile"
-	"github.com/docker/cli/cli/config/credentials"
-	"github.com/docker/cli/cli/config/types"
-	"oras.land/oras-go/v2/registry/remote/auth"
+	credentials "github.com/oras-project/oras-credentials-go"
```
Comment (shizhMSFT) [inline internal/credential/store.go:19]: Is this `credentials` alias auto generated by the go tools?

File: cmd/oras/root/login.go (hunk 1 above)
Comment (shizhMSFT) [reply cmd/oras/root/login.go:111]: Let's remove this comment.

**Review by shizhMSFT — COMMENTED** [review]
-

File: internal/credential/store.go (hunk 4 above)
Comment (qweeah) [reply internal/credential/store.go:19]: yes

File: internal/credential/store.go [hunk 5]
```diff
@@ -20,21 +20,23 @@ import (
...
-func NewStore(configPaths ...string) (store credentials.Store, err error) {
-	so := credentials.StoreOptions{AllowPlaintextPut: true}
+func NewStore(configPaths ...string) (credentials.Store, error) {
+	opts := credentials.StoreOptions{AllowPlaintextPut: true}
+	var store credentials.Store
+	var err error
 	if len(configPaths) == 0 {
 		// use default docker config file path
-		store, err = credentials.NewStoreFromDocker(so)
+		store, err = credentials.NewStoreFromDocker(opts)
```
Comment (shizhMSFT) [inline internal/credential/store.go:29]: How about
```suggestion
		return credentials.NewStoreFromDocker(opts)
```

File: internal/credential/store.go [hunk 6]
```diff
@@ -20,21 +20,23 @@ import (
...
 	} else {
 		var stores []credentials.Store
```
Comment (shizhMSFT) [inline internal/credential/store.go]: Then we can do
```suggestion
	}

	var stores []credentials.Store
```

File: internal/credential/store.go [hunk 7]
```diff
@@ -20,21 +20,23 @@ import (
...
 		for _, config := range configPaths {
-			store, err := credentials.NewStore(config, so)
+			store, err := credentials.NewStore(config, opts)
 			if err != nil {
 				return nil, err
 			}
 			stores = append(stores, store)
 		}
 		store, err = credentials.NewStoreWithFallbacks(stores[0], stores[1:]...), nil
```
Comment (shizhMSFT) [inline internal/credential/store.go:39]: And then
```suggestion
		return credentials.NewStoreWithFallbacks(stores[0], stores[1:]...), nil
```

File: go.sum [hunk 8]
```diff
@@ -2,14 +2,8 @@ github.com/cpuguy83/go-md2man/v2 v2.0.2/go.mod h1:tgQtvFlXSQOSOSIRvRPT7W67SCa46t
 github.com/davecgh/go-spew v1.1.0/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
 github.com/davecgh/go-spew v1.1.1 h1:vj9j/u1bqnvCEfJOwUhtlOARqs3+rkHYY13jYWTU97c=
 github.com/davecgh/go-spew v1.1.1/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
```
Comment (TerryHowe) [inline go.sum:4]: A lot of dependencies cleaned up here.

**Review by TerryHowe — APPROVED** [review]
LGTM

File: go.sum (hunk 8 above)
Comment (qweeah) [reply go.sum:4]: This is expected after `go mod tidy`

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**